src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs

```diff
@@ -5349,5 +5349,39 @@ static void Main()
             string expectedOutput = @"1";
             CompileAndVerify(source, expectedOutput: expectedOutput);
         }
+
+
+        [Fact]
+        public void TestNullCoalesceWithDefault()
+        {
+            var source = @"
+class Program
+{
+    static int CoalesceInt32(int? x)
+    {
+        return x ?? 0;
+    }
+
+    static T CoalesceGeneric<T>(T? x) where T : struct
+    {
+        return x ?? default(T);
+    }
+}";
+            var comp = CompileAndVerify(source);
+            comp.VerifyIL("Program.CoalesceInt32", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""int int?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+            comp.VerifyIL("Program.CoalesceGeneric<T>", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""T T?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+        }
```

Comment:
>} [](start = 8, length = 1)

It would be good to have a test for a cases:
- when the right side is not nullable and not default
- when the right side is nullable and not default 
- when the right side is nullable and default

Reply:
- when the left side is a reference type, the right side is default 

---
In reply to: [194488414](https://github.com/dotnet/roslyn/pull/27700#discussion_r194488414) [](ancestors = 194488414)

Reply:
I would also add a test for when `GetDefaultValue` is missing. I think you should be able to use `compilation.MakeMemberMissing` for that.

Reply:
I've added the tests.

For `GetValueOrDefault`, I've also added the test but it feels redundant with the `MissingSpecialMember` tests. I've tried to use `Compilation.GetSpecialMember` to avoid the diagnostic and fall back to the unoptimized code generation, but it still needs `GetValueOrDefault` anyways so in the end I kept the call to `TryGetNullableMethod`.

Reply:
> For GetValueOrDefault, I've also added the test but it feels redundant with the MissingSpecialMember tests. I've tried to use Compilation.GetSpecialMember to avoid the diagnostic and fall back to the unoptimized code generation, but it still needs GetValueOrDefault anyways so in the end I kept the call to TryGetNullableMethod.


The point of this test is to ensure the right diagnostics is reported and compiler does not crash or fail in other unexpected ways.

---
In reply to: [195272843](https://github.com/dotnet/roslyn/pull/27700#discussion_r195272843) [](ancestors = 195272843)